feat: add plugin compatibility checks and override UI

OGKevin opened the pull request:
This pull request introduces a new plugin compatibility checking subsystem for KOReader, allowing plugins to declare compatible version ranges and providing users with options to override compatibility decisions. The implementation is configurable via a new setting and includes both documentation and a runtime system for managing plugin compatibility and user overrides.

**Plugin compatibility subsystem:**

* Added a new setting `ENABLE_PLUGIN_COMPATIBILITY_CHECKS` in `defaults.lua` to toggle plugin compatibility checks based on declared version ranges in each plugin's `_meta.lua`.
* Implemented the core compatibility checking and override management logic in the new file `frontend/plugincompatibility.lua`, including UI integration for user override options and persistent settings.

**Documentation:**

* Added `doc/plugin_compatibility.md` to explain how plugin compatibility works, how to declare version ranges, available user override actions, and when to use compatibility constraints.

Fixes: https://github.com/koreader/koreader/issues/14613

<details><summary>Screenshots</summary>

<img width="540" height="720" alt="2025-11-29-112109_hyprshot" src="https://github.com/user-attachments/assets/21744543-5196-49b6-986f-d763c9f40977" />

<div style="display:flex; gap:8px; align-items:flex-start;">
Before:
  <img alt="2025-11-29-112118_hyprshot" src="https://github.com/user-attachments/assets/51afa4a7-adbc-4dcf-8e57-9e82b4c8ea5f" width="300"/>
After:
  <img alt="image" src="https://github.com/user-attachments/assets/556aa829-2db7-4d48-942f-8689da218019" width="300"  />

</div>



<img width="540" height="720" alt="2025-11-29-112308_hyprshot" src="https://github.com/user-attachments/assets/a4ab4ad3-2361-434f-a80b-6f49d428fe9f" />

<img width="540" height="720" alt="2025-11-29-112126_hyprshot" src="https://github.com/user-attachments/assets/fd4985bd-2c1a-46a1-b3cc-ba49cc73e3d8" />

<img width="540" height="720" alt="2025-11-29-112400_hyprshot" src="https://github.com/user-attachments/assets/38707ffd-3a8a-4cde-bfc6-d5b00e5e4a11" />

<img width="540" height="720" alt="2025-11-29-112809_hyprshot" src="https://github.com/user-attachments/assets/b605e32d-7798-4673-812d-358a24bb23ee" />


<div style="display:flex; gap:8px; align-items:flex-start;">
Before:
 <img alt="2025-11-29-112813_hyprshot" src="https://github.com/user-attachments/assets/47041e3e-b068-4915-ad39-feed8f66f23c" width="300"/>
After:
 <img  alt="2025-11-30-111908_hyprshot" src="https://github.com/user-attachments/assets/359806c1-eb9a-4c1e-a27a-56fa3263f4e1"  width="300" />
</div>
</details> 

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/14655)
<!-- Reviewable:end -->

Frenzie commented on frontend/plugincompatibility.lua:
```diff
@@ -0,0 +1,423 @@
+
+    local sub_menu = {
+        {
+            text = _("Incompatibility Details"),
```
```suggestion
            text = _("Incompatibility details"),
```

Frenzie commented on frontend/plugincompatibility.lua:
```diff
@@ -0,0 +1,423 @@
+
+    local main_menu
+    main_menu = Menu:new {
+        title = _("Incompatible Plugins"),
```
```suggestion
        title = _("Incompatible plugins"),
```

Frenzie reviewed (commented):
The overall structure looks decent to me at a a glance; I didn't look at the details. I left a couple of minor comments about strings/localization.

joshuacant commented:
> But why force these developpers, with a max stuck at the current KOReader version, to have their plugin DISABLED when the user will update KOReader, while that plugin may very likely still work... Feels like provider (the developper of that external plugin) driven bad user experience: you updated, you don't get my feature anymore, until I developper find the time to bump the max version (because he will probably not need to do anything more) - and you user find the time to (find again and) update that plugin...

I want this feature very much.

I'm a terrible, amateur programmer and I consider the effect my code may have on people's devices twice as seriously as a result. If I wrote a plugin that is manually accessed by a menu and it crashes, then no big deal. KOReader reloads and the user knows not to do that again and seek out help to resolve the crash, either by filing an issue or updating or more likely filing an issue and being told they need to update. 😜 But they can still access their library, their books. This is not so for plugins that load at launch or immediately upon opening a document.

The plugin I have written loads at launch. If it crashes, the device is sent into a crash loop and the user is completely prevented from getting back into KOReader until they return to a computer, plug their device in, and figure out what's going on. Roughly half of the updates to KOReader since I first shared Project: Title have resulted in this. The other half have had crashes around specific functions which are less bad, but still very annoying to the user.

The scenario I consider is someone who loaded up their device with books for a long flight or vacation and then, at the last minute, decides to update KOReader from the airport or hotel WiFi. Now their device crash loops and they are prevented from doing anything with it until they can get back to a computer and spend time fixing it. Is this unlikely? Sure. Should I still take steps to prevent it from happening? I think so, yes.

I already wrote code to check the KOReader version number and load a dummy/blank plugin with an error message if the version doesn't match. I would still prefer seeing that concept formally embraced and handled by KOReader itself. Given that I had nothing to do with filing the issue or opening this PR, it seems at least one other plugin developer does too. 😄 

When I decided to publicly share what I was making, an explicit, stated goal of mine was to do what I could with what KOReader provides, not file any issues or requests, and try my absolute hardest to prevent any extra work for you and your team. I haven't been entirely successful in the latter, and I deeply regret the issues people have filed here that were reporting mistakes I made. I'm making an exception to the former by advocating for this PR because I feel like my perspective may be of use. However I understand completely if this PR is not approved and you will never see me say a bad word about the KOReader team. Project: Title is nothing more than a different style of wallpaper for a house that I couldn't even dream of building.

> (Looks like there is a more active life and experiments out there :) https://www.reddit.com/r/koreader/comments/1p433fa/app_store_plugin_for_koreader/)

This is 6,000 lines of un-reviewed LLM code. It is not known yet if all parts of the code function or if it will be maintained frequently in the future.  I have not looked at the code or documentation, so I don't know if it even has min/max version checks. There is one other plugin out there right now that offers in-place online updates for plugins and patches, and it is also thousands of lines (>4,000) of LLM code and also has no version check in place. As a result, I do not see either as a replacement for what this PR provides, rather it is complementary.

Anyway, I apologize for the length of this comment and appreciate the time anyone took to read it. Thank you for everything anyone on your team does, big or small, to keep KOReader working on so many devices.

joshuacant commented:
> Our plugin loader that is supposed to pcall-load the plugin does not prevent that ? Could we do something ?

This is above my skill level with Lua, I fear. I can provide details (somewhere else?) on the crashes I've caused as a result of upstream changes if that would be of interest. IMO even better than a min/max version check would be a way for KOReader to prevent plugin misbehavior or identify a misbehaving plugin and disable it, thereby preventing inescapable crash loops.

> I glance every other day at the koreader reddit, I was just mentionning this thread among others (and we probably have an issue opened about plugin management/updates). There are lots of ideas and experiments, but I'm glad it happens elsewhere. For most of these stuff, if it would land here or if we ourselves would have that idea, we'd probably spend weeks discussing every little bit or aesthetics/UI decisions, allowing for flexibility/options via menu, but not too many (my obsession). I haven't followed how you bear that with your user requests :)

I too started checking the koreader subreddit every few days, after realizing how much discussion about plugins was happening there. :)

When I was working on my plugin, I realized how much effort your team has put into allowing plugins and patches and even patches that target plugins. I interpreted the unspoken message as, "please use this functionality rather than bringing everything to us."

I took that philosophy to heart when making my plugin and have embraced it when it comes to the feature requests I get. When I get a request I don't want to add I'll think through how hard it would be to write a user patch that does it. Often times there are simple changes I can implement that make a user patch trivial and so I will do that instead.

OGKevin commented:
Don't take this the wrong way, but there is a reason I've written a detailed proposal with implementation direction and have left it open for ~10 days to give anyone interested a chance to provide feedback. 

You haven't provided any feedback on it, and now you place comments against it + asking questions that have been answered in it. This could have been discussed before this PR was submitted. 

---

Judging an implementation purely on loc is meaningless. I can print hello world with 1000 useless loc or a fancy unreadable oneliner. I suggest to just review the code when its in a reviewable state and make proper judgements. 

An overly documented solution with enough guardrails and putting the user in control, will take some loc. Especially if we make it so that its is fine grained per plugin and KOReader version combination. 

Your suggestion could as well be 423 loc 🤷🏾‍♂️ , again, on it's own its meaningless. 

> And the docstrings, paradoxically, don't help a quick reading with their tons ot @tparam that state the nearly obvious from the parameter name...

What is obvious for you, might not be obvious for others. Or 10 years later when someone new tries to understand the code. But if you insist, they can be removed. 
Lua is not a strictly typed language, so you can understand how one would appreciate type hints.  

Overengineering is subjective, people with different engineering backgrounds have different opinions on what this means. For someone with my background and the problem this is trying to solve, it doesn't look overengineered.

Concrete feedback on what you feel as overengineered would be better feedback. Most likely I would be able to answer why I've done things the way I've done it. And out of that discussion we can come to some conclusion.

---

> Btw, what if the user updates to a nightly, load once, check all is good, then says "load always". And then next week updates to another nightly, where the plugin will break? Will he be asked again on each KOReader update ?

Yes, but if this is undesired, an alternative can be discussed. e.g. disable all compatibility checks because you're an advanced user that runs bleeding edge. They key target audience are the everyday non tech savvy user who hits OTA for KOReader and lands in a crash loop.

---

I suggest you folks, KOReader maintainers, decide how to move forward. The problem statement outlined in the proposal is clear, and this is one way to implement a solution. You must decide if this is the way you want to see it be done, or have it done any other way.

poire-z commented:
You're right, sorry for not having taken a better interest in your proposal. I did look at it, it felt it wasn't really my topic of interest, and the four 5-30 lines snippets felt it wouldn't hurt :)
Now that I see its realization in this PR, and the amount of code, I had to have a deeper look and thinking.

I'd like other maintainers to join and force themselves to think too :) but in the meantime, please bear with @Frenzie and me :)

Also, about line of codes and overengineering, dont take this the wrong way too :):
I think I mentionned this in your previous PR, that didn't end up and became your comicreader plugin.
You seem to come from a more professional-coding background than most of us: your unit tests, your use of docstring for most functions, your split of code in multiple little functions, your use of a lot of blank lines (which means you must have a large and possibly vertical monitor :)), your use of an IDE and your complain of the lack of style hints(?) in our code base and us not caring in your other PR.
All this is fine, and good for you.
But the state of KOReader up to this day is that it has been made by non-professionals with simple text editors copying and pasting and adapting bit from elsewhere in the code base. I think we (at least I) like it concise without noise and easy to read (small laptop screen, even reviewing this PR on github web have me scrolling all around). It may look odd and stupid to you, but it suits most of us well I think (I know @Frenzie likes docstrings and unit tests :) the sample above for RenderText was probably forced upon me to write it by him :)). I also feel that this has somehow becoming our common code style, so anything different comes up as alien and looks odd, so my failed resistance to not mentionning that :/.
When I hack code around, having no style of my own, I just adapt to the surrounding code so my addition does not look too different. I may be abusing newcomers expecting them to behave the same way, but well, sorry. At least you can understand my point of view :)

Another point I made in your other PR (which apply a small bit only to this one) is invasiveness of "second-class" features into our existing first-class and difficule code. You didn't comment on this in that other PR - but if you had, I might have been fine with it: with the amount of work you had to do in your plugin to plug/patch into core your stuff, you would only have need to do half of that in the other PR to have it look fine to me (half of that because you would have been allowed to add function calls into our core functions.

More practical comments and thoughts about the issue in my next comment.

Frenzie commented on frontend/plugincompatibility.lua:
```diff
@@ -0,0 +1,423 @@
+        -- User wants to load it once for testing
+        -- Clear the override so next time it won't auto-load
+        PluginCompatibility.clearLoadOnceOverride(G_reader_settings, plugin_name)
+        PluginCompatibility.removePromptedMark(G_reader_settings, plugin_name, plugin_version)
```
I do agree with @poire-z vis-à-vis <https://github.com/koreader/koreader/pull/14655#issuecomment-3572243283> that extracting a handful of meager lines like this into their entirely separate functions doesn't aid clarity. Presumably it's primarily done for the sake of being able to test it in isolation, but as a consequence instead of having ~10 clear lines, it's ~30 unclear "overengineered" lines that require hopping back and forth. It's a trade-off of course, and one that bothers me less than it does @poire-z, but I am in the same camp nonetheless.

Besides reduced maintainability aspect, it's also a mildly suspicious pattern in the sense that it would be worse in performance-sensitive code. As such doing so unnecessarily might be a bad habit even where it makes no difference.

OGKevin replied:
Testing is one key contributor for splitting the logic into smaller units. TDD (test driven development) is one way to get things done. This allows for one to no longer care about the internals of a function that is under test once it passes them.

Nevertheless, I am fine with merging things in bigger functions if that is the preference. I would need some pointers on what you would like to see merged. Since we see things differently. 
I am also happy to just deliver a working version and you folks can take it over from there and refactor it as needed.

poire-z commented:
I still think the idea of min and max version isn't really worth it. For the reasons explained above, the fact that min & max version will probably always be the same, the case of "always load"/advanced users that won't help on the next problematic update, etc... It's also a bit limited, and force the code to deal with just one (not really good) solution and one condition (the koreader version).

Alternative idea: just let plugins be allowed to have a `compatility_check(koreader_version)` function in their `_meta.lua`. We pass the koreader_version as a convenience for quick one-liners allowing what you wished for (validate version(s) with which it is compatible if one really wants that - but then the logic is in the plugin, not hardcoded into core. It would return true/false, and some informative message about how to handle the possible incompatibility and where to find updates. (All this at the discretion of the developper, we don't need to know.)
More adventurous developpers could do other checks, like checking for existence, or fetching the function from readerview and doing a sha1 on the function code text (we can get it, I did that in the httpinspector plugin), comparing it to a known sha1, and if identical, probably nothing has changed, no need to worry the user with a ⚠ or disable the thing.
Or do any other thing that make sense - without core needing to know or provide services (the function code extraction and sha1 computation could be provided by core in a compatibility.lua, like we do provide stuff in frontend/userpatch.lua).

And I don't think this need to be a hard rule, just an information.

I guess a plugin destiny after an update is one of these:
A) it just works
B) it failed to load plainly (without side effects like half-monkey-patching): the user may notice it's just not there, and go into the plugin management menu to see if it is enabled or not (do we notice and act on that by showing it disabled?). This could also be explicitely triggered by the author if he really wants the thing to not be enabled when the compatibility check did not validate).
C) it loads or half-loads, some/all features do not work, but it does not crash: the user may notice it and go see that it is enabled, and would see ⚠ (and get some author information about the possible incompatibility on long-press), and could disable it. In this case, can some things be really irrevocably messed up/destroyed ?
D) it crashes KOReader's startup and there's no way to disable it
E) any other situation I didn't imagine ?

I think your submenu appearing on "incompatible" plugins and the options of load-once/never/always are a bit unneccesary (and a bit overengineered, providing an UI we don't need) and don't really solve the issues with future updates after the user has made a choice.
(Just curious: do you think your users (most must be on reddit?) stick with a stable to just enjoy your stuff, or do they use nightlies to also enjoy _our_ stuff ? :))

A/B/C can be handled by the user and our regular plugin management menu checkboxes - with that ⚠ and some developper information on long-press.

For D, I guess we need a safe mode. If startup did not finish (flag stored somewhere (on the filesystem I guess?) at start, removed after readerready or filemanagerready, or a few seconds after that? (x)), on next startup if flag present, disable ALL user plugins and user-patches - and remove that flag. The user can then go re-enable them one by one.
((x): is it possible it could crash on the first menu opening (it has happened when developping core, dunno if it can in other contexts), and then the user could not go anywhere to disable stuff...)

It's possible I missed something, but it feels it is simpler and more versatile than sticking with the per-plugin submenu and once/always/never (the choice won't reset on next KOReader update or plugin update).

----
If others feel the current PR is good enough, and about invasiveness, I would request that some more logic you added to pluginloader be put into plugincompatibility.lua: only a few function calls in pluginloader, and the disabled plugin list could be stored in plugincompatibility as well as the nextTick(showSomething) - all the things about incompatibility may not really need to add noise to the already complex pluginloader.lua.

hius07 commented:
The loaded_successfully flag looks simpler, and it handles user patches.
The downside: reenabling **all** user patches and plugins one by one (with restart) is not user friendly.

OGKevin commented:
> I think your submenu for marking plugins as "incompatible" and the load once/never/always options are a bit unnecessary, and somewhat overengineered, providing a UI we do not need. They also do not really solve the issues that can arise from future updates after the user has made a choice.

I do not follow. Can you elaborate on the "___do not really solve the issues with future updates after the user has made a choice___" bit?

Do you mean something like this?

> It is possible I missed something, but it feels simpler and more versatile than keeping a per plugin submenu with once/always/never: the choice will not reset on the next KOReader update or plugin update.

If so, then I think you are incorrect. The current implementation stores the decision like this:

```lua
plugin_compatibility_prompts_shown = {
  ["plugin_name_version_koreader_version"] = true,
}
```

That means if any of those three variables(plugin name, plugin version, koreader version) change, the user will be prompted again about what they want to do when incompatibility is not resolved. For bleeding edge users, they can disable compatibility checks altogether and always load all plugins.

Also, I would rather the system not fail silently. Providing a UI that indicates loading failed and lets the user take action is much better user experience than having a plugin silently fail and the user be confused why a feature they expect to work is not working.

About this idea:

> Alternative idea: just let plugins be allowed to have a compatibility_check(koreader_version) function in their _meta.lua. We pass the koreader_version as a convenience for quick one liners allowing what you wished for (validate version(s) with which it is compatible if one really wants that, but then the logic is in the plugin, not hardcoded into core).

This suggestion is not bad, and I would also advocate for it if we had a static language where we could require plugin authors to implement a specific interface and have the loader interact with it. The issue I see with this approach in Lua is that a bug, more specifically a nil pointer error, in `compatibility_check` can crash things. One drawback of interpreted languages is that it is risky to offload logic to external code.

There are two key things:
1. The actual compatibility check.
2. How to deal with incompatibility.

In my opinion, the second item should be handled by KOReader, not by plugins. That ensures consistent behavior when incompatibility occurs. If each plugin developer handled it differently: one using a popup, another silently disabling the plugin, another using a full screen menu, the user experience would be inconsistent and potentially chaotic, especially if multiple failures and UI widgets stack.

So your `compatibility_check` suggestion could be useful if it only returned a boolean indicating whether the plugin should be loaded. But I still worry: in Lua it is hard to enforce that external code follows the expected contract, for example returning values like `shouldLoad` and `reasonForNotLoad`. I believe it is safer and simpler to keep `_meta.lua` focused on metadata about the plugin.

---

But like I said, it's up to you folks to make the final decision. 
I like the idea of being prompted when incompatibility is detected, and the safety of being in control of how this check is done. Especially if you want/plan to move/use this for full fledged "safe mode". In that case, I would argue that KOReader must be in full control and external plugins shouldn't dictate behaviour. e.g. `compatibility_check` raising some popup/notification.

poire-z commented:
> This suggestion is not bad, and I would also advocate for it if we had a static language where we could require plugin authors to implement a specific interface and have the loader interact with it. The issue I see with this approach in Lua is that a bug, more specifically a nil pointer error, in `compatibility_check` can crash things. One drawback of interpreted languages is that it is risky to offload logic to external code.

Sure. But we're using Lua, we are lazy, we make a reader app, we are not enforcing any API, everything in the code is accessible to be hacked and monkey patched. (What users, external user patch and plugins developpers, and their users, do is not our problem - we don't want this to be another constraint on us. But I'm fine with providing hooks and simple ways to manage the issues they may have).

As far as a compatibility_check() function is offered to plugin authors, to be simple and only returing a boolean and a message, we can assume the author will have tested it once :) and it has no reason to crash on a KOReader update. In this issue, we are only dealing with KOReader updates, not about the quality of plugins.
On our side, the loader just have to check if this compatibility_check property exists, that it is a function, and call it: it gets nil, or false, or anything else that is considered true (and a string or nit as the 2nd returned value), the loader acts on that, nothing more is expected from the plugin.


> There are two key things:
> 
>     1. The actual compatibility check.
> 
>     2. How to deal with incompatibility.
> 
> In my opinion, the second item should be handled by KOReader, not by plugins. That ensures consistent behavior when incompatibility occurs. If each plugin developer handled it differently: one using a popup, another silently disabling the plugin, another using a full screen menu, the user experience would be inconsistent and potentially chaotic, especially if multiple failures and UI widgets stack.

I'm with you. I didn't expect the plugin to do the notifications: compatibility_check just returns a boolean and a string. It's the loader that would accumulate the results and messages, and display them (or not).

> I do not follow. Can you elaborate on the "_**do not really solve the issues with future updates after the user has made a choice**_" bit?
> 
> Do you mean something like this?
> 
> > It is possible I missed something, but it feels simpler and more versatile than keeping a per plugin submenu with once/always/never: the choice will not reset on the next KOReader update or plugin update.
> 
> If so, then I think you are incorrect. The current implementation stores the decision like this:
> 
> ```lua
> plugin_compatibility_prompts_shown = {
>   ["plugin_name_version_koreader_version"] = true,
> }
> ```
> That means if any of those three variables(plugin name, plugin version, koreader version) change, the user will be prompted again about what they want to do when incompatibility is not resolved. For bleeding edge users, they can disable compatibility checks altogether and always load all plugins.

OK, I didn't notice the koreader version on which the choice has been asked is stored - and the choice applies only for this version.
So, on any next nightly update, the user gets again in the process of ack'ing all plugins ? Feels it is a bit too much, and bothering.
 
> Also, I would rather the system not fail silently. Providing a UI that indicates loading failed and lets the user take action is much better user experience than having a plugin silently fail and the user be confused why a feature they expect to work is not working.

This is where we differ: most of the times, things won't fail. 
You think the compatibility check is a hard rule: if it returns false, we don't load at all: it's for the user to go thru dialogs/menus  to accept it and restart KOReader (because we still need to restart for enabling a plugin after all are loaded, right ? Or do you ask questions and block for answers while still in the plugin loading phase?).
I think it should not be a hard rule and not block anything. Most of the time, this is the right decision, even for your heavy plugins (we rarely touch CoverBrowser and ReaderView, so it's possible they won't be changed even between a stable and the next). The result of compatibility_check may just have us put this ⚠ in the menu item for that plugin, so the user can see where he may fix stuff or find infos. We may auto-disable it if we detect the plugin loading failed. The user will know that (and you could tell it after the loading phase with the accumulated checks and strings).
(The possible crash loop is another issue that would not be fixed by this compatibilty_check thingie, but by another PR/feature: that safe mode.)

Our differences all come down to you being driven by static typing, api, contracts, user ~~harassment~~prompts :) And KOReader (not me) being driven by freedom, careless natural readable simple code, free access to internals, user-patch, do-the-preferable-thing by default and allow users to tweak that later; in a word, by love :)

> But like I said, it's up to you folks to make the final decision.

Well, we like to discuss a lot :) Sometimes, some clever idea come up and we may all agree on it.

I insist on expressing all this not to harrass you, but so you can understand better KOReader's philosophy.
This PR might be merged if I did not manage to convince you about some alternative and if other maintainers are just fine with it. I will just request the plug into pluginloader to be more minimal, and as far as what happens in compatibility.lua (the test tricks, G_reader_settings, the docstrings overload, the blank lines, the multiple little functions), I can forget them as I'll probably never open this file. (We already have a demi-dozen of such files, made by one-time contributors.)
But I'd like you to get the overall idea of how we like things. If in your next PR you touch more core stuff with all the same things, I'll probably just repeat all that - and we'll both be tired :)

OGKevin commented on frontend/pluginloader.lua:
```diff
@@ -186,45 +188,110 @@ function PluginLoader:_discover()
+
+        -- Check compatibility before loading the main plugin file
+        local should_load, incompat_reason, incompat_message, should_prompt =
+            PluginCompatibility.shouldLoadPlugin(G_reader_settings, plugin_meta, plugin_name)
```
> The good news is this discussion may have uncovered a bug. :saveSetting() works in memory. If you want to write to disk you should call :flushSetting() or :close().

ah nice spot 👍🏾

OGKevin commented on frontend/plugincompatibility.lua:
```diff
@@ -0,0 +1,458 @@
+
+-- TODO: there must be a way to clean up lingering settings.
+-- e.g. when a plugin is uninstalled, or when a new KOReader version is released
+-- that makes old overrides irrelevant.
```
How would you imagine this to be solved @Frenzie ? 

Once per plugin load cycle, check for "garbage" settings entries and delete them? But keep `n+2` where `n` is current version?

Frenzie replied:
I'm not entirely sure if I understand what you mean by +2 in this context but in essence sure.

OGKevin commented on frontend/pluginloader.lua:
```diff
@@ -192,39 +204,119 @@ function PluginLoader:_load(t)
+    self:getCompatibility():showIncompatiblePluginsMenu(incompatible_plugins, function()
+        UIManager:askForRestart()
+    end)
+end
```
```suggestion
    -- If there are incompatible plugins that need prompting, show the menu
    if #incompatible_plugins_with_prompt > 0 then
        UIManager:nextTick(function()
            self:getCompatibility():showIncompatiblePluginsMenu(incompatible_plugins_with_prompt, function()
                UIManager:askForRestart()
            end)
        end)
    end
end
```

OGKevin commented on frontend/pluginloader.lua:
```diff
@@ -192,39 +204,119 @@ function PluginLoader:_load(t)
+    local incompatible_plugins_with_prompt = {}
+    local compatibility = self:getCompatibility()
+
+    for id, v in ipairs(t) do
```
using `_` will clash with `local _ = require("gettext")`

```suggestion
    for __, v in ipairs(t) do
```

hius07 commented:
UI suggestions:
-1st screenshot: remove the submenu marker after the plugin name
-2nd screenshot: use ButtonDialog instead of this Menu widget
-7th screenshot: use radiomarks

Frenzie commented on frontend/plugincompatibility.lua:
```diff
@@ -0,0 +1,367 @@
+local UIManager = require("ui/uimanager")
+local Version = require("frontend/version")
+local _ = require("gettext")
+local logger = require("logger")
```
Please abide by https://github.com/koreader/koreader-base/wiki/Coding-style#order-of-requires

Frenzie commented on frontend/plugincompatibility.lua:
```diff
@@ -0,0 +1,367 @@
+    local max_version = compatibility.max_version
+    -- Check minimum version requirement
+    if min_version then
+        local min_ver, __ = Version:getNormalizedVersion(min_version) -- luacheck: ignore
```
If you're not using it you can just leave it off.

Frenzie commented on plugins/hello.koplugin/_meta.lua:
```diff
@@ -1,6 +1,14 @@
+if os.getenv("KODEV_ENABLE_INCOMPATIBLE_PLUGIN") then
+    meta.compatibility = {
+        min_version = "v0000.01-1",
+        max_version = "v1111.09-9",
```
Shouldn't that be 9999?

Frenzie replied:
I'd just reduce it to KODEV_INCOMPATIBLE_PLUGIN. The "enable" is both redundant and confusing.

Frenzie commented on defaults.lua:
```diff
@@ -197,4 +197,14 @@ DUSE_TURBO_LIB = false,
+-- to override the decision (ask on incompatibility, load once, always load,
+-- or never load). Disabling this flag bypasses those checks and allows all
+-- plugins to be loaded regardless of their declared compatibility.
+-- Note: the runtime implementation lives in `frontend/plugincompatibility.lua`.
```
Settings documentation isn't meant to be developer documentation. :-) Let's tone it down a bit in this general direction:

```suggestion
-- When enabled, each plugin's declared compatibility will be validated
-- against the running KOReader version.
-- If a plugin is incompatible, loading is blocked and the user will be prompted
--  (ask on incompatibility, load once, always load, or never load).
-- Disabling this flag bypasses those checks and loads all plugins regardless of compatibility.
```

Frenzie commented on spec/unit/plugincompatibility_spec.lua:
```diff
@@ -0,0 +1,322 @@
+        -- Reset the settings data to ensure clean state
+        compatibility.settings:reset({
+            plugin_load_overrides = {},
+            plugin_compatibility_prompts_shown = {},
```
This is one of the first useful things I've seen come out of a Copilot review. :-)

Frenzie commented:
As previously stated I'm basically fine with the code flow, and after the latest changes I like the way the tests and docstrings look now too. I might have a few minor UI and string comments but I think I'll wait until the Copilot-typos are resolved to take a more detailed look at that.

OGKevin commented:
I believe I've addressed all the comments.

@hius07 I've updated the screenshots as well 🤝. 

I also had to add 2 new more tests for a bug: ___showIncompatiblePluginsMenu and ButtonDialog callback behavior___

poire-z commented:
> _If others feel the current PR is good enough_, and about invasiveness, I would request that some more logic you added to pluginloader be put into plugincompatibility.lua: only a few function calls in pluginloader, and the disabled plugin list could be stored in plugincompatibility as well as the nextTick(showSomething) - all the things about incompatibility may not really need to add noise to the already complex pluginloader.lua.

As this first _if_ seems to be true :(, thoughts about my _then_ request ?
You also removed a lot of blank lines in your modules (thanks), but there are still a lot that remains in pluginloader.lua.

Frenzie commented:
Yes, that's roughly what I mean. As stated before I find it acceptable and desirable. I'll just need to find the time to look over the pluginloader changes more closely.

poire-z commented on frontend/pluginloader.lua:
```diff
@@ -192,40 +204,116 @@ function PluginLoader:_load(t)
+                incompatibility_message = incompat_message,
+                should_prompt_user = should_prompt,
+                version = plugin_meta.version,
+                fullname = plugin_meta and plugin_meta.fullname or plugin_name,
```
You mean:
```lua
local should_prompt, plugin_stub = compatibility:shouldLoadPlugin(plugin_meta)
if not should_prompt and not disabled then
  -- our original loading code
```
?
I don't really mind the naming (although it should read like what the function name means, not the opposite :) unlike `dont_load = compatibility:shouldLoad()` which `should_prompt` sounds like).
What I mind is having only one branch, with the same inner code we had before, you would just update the condition.
And what I mind more is having the minimal stuff related to compatibilty in the existing pluginloader.lua.

OGKevin replied:
Right. I'll make some change.

If afterwards you still have feedback, I kinldy ask for you to actually look at the code and provide concrete feedback that doesn't leave room for assumptions and interpertations from my side. This will most likekly trigger yet more comments, as we have completely different tastes and styles. 

I haven't moved anything from pluginloader yet, as I am waiting on some concrete feedback points from you and @Frenzie. So that I can address everything in 1 go. Unless the refactoring of pluingloader is the only standing feedback?

Frenzie replied:
> approach with min/max version, that I find it naive and unpractical, for users and developpers

You may it have mentioned it already, but what kind of alternative do you have in mind? Max version is inapplicable to perhaps every plugin I've ever written (it might occasionally apply many years down the line), but min version definitely is.

OGKevin commented:
In theory the code could be refactored (and improved, time complexity wise), by doing discovery and loading in one loop instead of 2. And than you won't need the ` mainfile = metafile` "hack". As you discover, and load it in place. But I think this would be for a diff PR tho, if you want to go in that direction.

poire-z commented on frontend/pluginloader.lua:
```diff
@@ -124,8 +126,18 @@ local PluginLoader = {
+        self.compatibility = PluginCompatibility:new()
+    end
+    return self.compatibility
+end
```
Is this needed ? What is th lifecycle of a PluginCompatibility instance ?
Our PluginLoader is (I think) a singleton: it's created at KOReader startup and the stuff it computed the first time plugins were discovered/loaded cached.
(And our singletons/modules that we instantiate at startup usually don't need to have a new(), we just use the class object the module has created and returned).
Can't we just have
`local PluginLoader = { ..., compatibility = PluginCompatibility }`
or if you really want to have a new():
`local PluginLoader = { ..., compatibility = PluginCompatibility:new() }`

poire-z replied:
Not really caring about one or the other - but why did you settle on using :new() ?
(And it's all used once, you could have spared the compatibility:reset() I think, because nothing would have used it before, no?)

poire-z commented on frontend/pluginloader.lua:
```diff
@@ -192,40 +204,74 @@ function PluginLoader:_load(t)
+    -- Show incompatible plugins menu if needed
+    compatibility:promptUserInNextTickIfNeeded(function()
+        UIManager:askForRestart()
+    end)
```
Feels (with my principles I expressed earlier :)) that we don't really need to know/read here all this about PluginCompatibility.
Something like just calling PluginCompatibility:finalize(), or:
```lua
-- Save settings and and prompt user:
compatibility:promptUserInNextTickIfNeeded()
```
with the askForRestart in that function.

poire-z replied:
> I do believe its important to indicate why saving in this particular place is important, and feared that if this is also abstracted away into the next function, it might be less obvious.

You care about this implementation detail because you've been swimming in it for weeks.
But in the context of pluginloader.lua, we don't need to know :) a `:finalize()` abstracting all it does would be just fine (and you would add comment in it that "pluginloader has loaded all plugins, now we can..." and what needs to be done, in this order, and why).

But ok if you really want 2 functions, just have nice names to read.

> What I can do is add a test case, that ensures that load-once plugins are only loaded once, and use that as "documentation". 

Not sure I understand - but again, in pluginloader, we don't need to know about your devel history, the bugs you met, any load-once issue, etc...

> it can be abstracted away as well.

Nice to hear :)

Frenzie replied:
> Move the inline doc into the abstracted function documentation, but this is a bit obscure and can lead to the same issue I'm trying to prevent.

That has been the suggestion from day 1.

poire-z commented on frontend/pluginloader.lua:
```diff
@@ -192,40 +204,74 @@ function PluginLoader:_load(t)
+        end
+        if not plugin_meta.version or plugin_meta.version == "" then
+            plugin_meta.version = "unknown"
+        end
```
Feels like you don't need the first part, and can just prepend `if not ok_meta or ..` to the 2 next parts as you are setting the same exact things as in the first part (so, it avoids typo not setting the same things).

OGKevin commented on frontend/pluginloader.lua:
```diff
@@ -192,40 +204,74 @@ function PluginLoader:_load(t)
-                    for k, module in pairs(plugin_metamodule) do
-                        plugin_module[k] = module
+        -- First load the metadata to check compatibility
+        local ok_meta, plugin_metamodule = pcall(dofile, metafile)
```
MR is Gitlab lingo indeed 🙈 

I'm referring to this: https://github.com/koreader/koreader/blob/7353a2e5a4fe28a0f7eaa879c0ed4cb5f7151d98/plugins/hello.koplugin/main.lua#L7-L10

Due to the hack, actual `main.lua` files can mark the plugin as disabled.
Changing this behavior without some investigations or discussion etc is unwise imo.
Hence I'm suggesting to do this separately. 

If you still think it shall be changed in this PR, it can be done ofc, but I advice against it 😛.

poire-z replied:
> Due to the hack, actual `main.lua` files can mark the plugin as disabled.

Yes, but I believe there are 2 different "disabled".
The one the user has set, which ends up as v.disabled in discover(), which is the one you could handle early.
And the one you mentionned that is handled by the `elseif type(plugin_module.disabled) ~= "boolean" or not plugin_module.disabled then` after we dofile(mainfile).
Isn't it?

Or are we talking about a different issue and about these 2 loops ?

poire-z commented on frontend/pluginloader.lua:
```diff
@@ -192,40 +196,60 @@ function PluginLoader:_load(t)
+        if not ok_meta or (not plugin_meta.name or plugin_meta.name == "") then
+            plugin_meta.name = plugin_name
+        end
+        if not ok_meta or (not plugin_meta.version or plugin_meta.version == "") then
```
Are the parens really needed ?

poire-z commented on frontend/pluginloader.lua:
```diff
@@ -255,12 +279,19 @@ function PluginLoader:genPluginManagerSubItem()
+            if plugin.incompatible and plugin.incompatibility_message then
+                element.fullname = element.fullname .. " ⚠"
+                element.description = element.description .. "\n\n" .. plugin.incompatibility_message
+            end
```
Feels like these tweaks could also be done by your `self.compatibility:tweakMenuItemIfIncompatible(menu_item, plugin)` ?

poire-z commented on frontend/pluginloader.lua:
```diff
@@ -192,40 +194,60 @@ function PluginLoader:_load(t)
     package.path = package_path
     package.cpath = package_cpath
-
+    self.compatibility:finalize()
```
You're allowed to add a small comment above it :)  so we don't have to go look at what it does, ie. `Save setting/state and prompt user if needed`.

pazos commented:
I'm sorry at being late to the party :/ I was busy in other projects :)

I like the idea of plugins restricting themselves to known versions but I feel this could be implemented as part of the plugin runtime.

For instance: https://github.com/koreader/koreader/blob/master/plugins/autodim.koplugin/main.lua#L9 showcases a plugin disabling itself based on device capabilities

So in that case the `compatibility` table is something internal to 3rd party plugins and pluginloader doesn't need to change.
 
Plugins can call `itWorks(t)` of some sort with min and max.

```lua

local compatibility = {
    min_version = "v0000.01-1",
    max_version = "v1111.09-9",
}

if not itWorks(compatibility) then
    return { disabled = true, }
end
```

If you think user feedback via UI is desiderable then ItWorks should display something when there's a mistmatch between runtime version and compatibility table

What do you think? Am I reinventing the wheel?

OGKevin commented:
@pazos see https://github.com/koreader/koreader/issues/14613#issuecomment-3703451931 and the linked comments. 

The problem of compatibility between 2 versioned components is a solved problem with battle tested solutions. 
Look at package managers, browser extensions, Obsidian plugins and so on.

> If you think user feedback via UI is desiderable then ItWorks should display something when there's a mistmatch between runtime version and compatibility table

See https://github.com/koreader/koreader/pull/14655#issuecomment-3576130467, specifically: 

> There are two key things:
> 
> 1. The actual compatibility check.
> 2. How to deal with incompatibility.
> 
> In my opinion, the second item should be handled by KOReader, not by plugins. That ensures consistent behavior when incompatibility occurs. If each plugin developer handled it differently: one using a popup, another silently disabling the plugin, another using a full screen menu, the user experience would be inconsistent and potentially chaotic, especially if multiple failures and UI widgets stack.

pazos commented:
lol, of course I missed the screenshots :sweat:

OGKevin commented on spec/unit/plugincompatibility_spec.lua:
```diff
@@ -0,0 +1,481 @@
+                name = "testplugin",
+                version = "1.0",
+                compatibility = {
+                    max_version = "2000.01-1",
```
Yea, these scenarios are iffy. Test passes because min_version is incorrect, which makes the check no-op, which marks it as compatible, which passes the tests. 

I'm inclined to add validation and error out if the value is not correct, would such a thing be acceptable @Frenzie? As the error should be seen so early in the dev cycle, I cannot imagine a dev adding an invalid compatibly, and not do a final test run and miss the crash.

OGKevin replied:
Essentially, what is the desired behavior on invalid input?

Frenzie replied:
Like I said, enforce a specific format. This should either already work or require only minor adjustment in Version. It's extremely unlikely that an LLM-style wrapper in plugincompatibility is the right solution.

Frenzie replied:
Yes. Just as a warning log though?

OGKevin replied:
Changed it to error 👍🏾

pazos commented:
Just my 2 cts again:

While having a table with a range of supported versions as part of the manifest is nothing bad by itself it introduces an exception on how we deal with plugins.

Until now plugins (and plugin developers) were used to check at runtime if a basic feature was available to them and decide based on that to fill the plugin table or disable the plugin all together.

And since we assume plugins decide we have a nifty `pluginloader` that doesn't try to act smart about loading plugins or not because it delegates that work to plugins themselves.

I wonder if it makes sense to introduce an exception and I wouldn't like to declare in the manifest: "requires: ambient-light" and the likes

Frenzie commented:
> We can comprimise, and have both solutions?

I wouldn't frame it as a compromise. Most plugins only need a minimum version, some need a maximum version, some need slightly more complex conditions.

Mind, you could still easily put most feature checks in a table form along the lines of:

```lua
-- probably also needs a way to say NOT canSuspend though
requires = {
 "canSuspend",
 "hasFrontlight",
 "isKobo"
}
```

```lua
for _, method in ipairs(requires) do
    if type(Device[method]) == "function" and Device[method]() then -- probably with pcall
        return true
    else
        return false
    end
end
```

> But yea, I'm not a huge fan of ofloading work to externals from Lua.

pcall exists; I think we'd still want more custom compatibility checks (e.g., binary x, y, z is not available in PATH) to be able to easily hook into the UI as well in some manner, probably just by requiring plugincompatibility. If it's simple to do you could add it to this PR, but for me that doesn't need to be done for this to be merged in any case.

poire-z commented:
> Now, for complexer scenarios, e.g:
>     * needs to be a kobo device
>     * needs to have frontlinght
>     * needs to have g sensor
>     * etc

That's device features, and I think it's orthogonal to what this PR handles: there's no reason to allow for "load once" "always load" when the device doesn't have the needed features.

> The addition of a should_load anonymous function in _meta.lua that returns true or false, where one can put these advanced checks, would work?
So if all you need is simple version constraint, you use what this MR proposes, if you need fancier/additional checks, you implement function that returns bool.

This, I'm fine with (and it's what I proposed in https://github.com/koreader/koreader/pull/14655#issuecomment-3572750795) - it's a more generic solution (that could just replace your min/max_version, without much more work needed for plugin authors, stripping this PR from half its unit tests about versions comparisons).

Frenzie commented:
In the context of the wider plugin ecosystem such metadata is also something that for example <https://github.com/omer-faruq/appstore.koplugin> could make use of. That's somewhat orthogonal to what pluginloader/plugincompatibility does, but if it were purely unused metadata the chance of being outdated also increases. :-)
